modules/internal/ArrayViewRankChange.chpl

```diff
@@ -22,9 +22,337 @@
 // This module provides an array domain map class that is used to
 // represent rank change slices of another array via a domain.
 //
+// Throughout this file, 'down*' refers to the higher-dimensional
+// array/domain/distribution that's been rank-changed in this view,
+// while 'up*' refers to the lower-dimensional, user-facing view.
+//
+// TODOs for this file:
+//
+// * the dist, dom, and arr classes store a certain amount of
+//   duplicated information -- should these be focused in the higher
+//   classes (dist/dom) and referenced by lower ones (dom/dist)?
+//
+// * I don't think I've done anything to privatize the dist or dom
+//   classes... should I?
```

Comment:
A comment higher in the function (on the Dom class) asserts that this only makes sense on rectangular.  I don't think rank change makes sense for non-rectangular domains/arrays because associative and opaque are arguably rank-1 (so rank-change could only reduce them to rank-0).

I think reindexing similarly doesn't make sense for non-rectangular because it would essentially require O(numIndices) space to store the map from old to new indices.

As you note, slicing makes sense for all three and I believe the slicing implementation doesn't do anything to prevent slicing a dense with a sparse, an associative with another associative, etc. (though I haven't tested many of those cases yet; though it's on my TODO list and the testing system suggests we at least haven't broken anything we'd verified as working).